src/main/java/ch/jalu/configme/resource/PropertyPathTraverser.java

```diff
@@ -1,78 +1,76 @@
-            parentPathElements, pathElements.subList(0, pathElements.size() - 1));
-        List<String> newPathParts = CollectionUtils.getRange(pathElements, commonPathParts.size());
-
-        parentPathElements = pathElements.subList(0, pathElements.size() - 1);
-
-        int indentationLevel = commonPathParts.size();
-        String prefix = commonPathParts.isEmpty() ? "" : String.join(".", commonPathParts) + ".";
-        return convertToPathElements(indentationLevel, prefix, newPathParts);
-    }
-
-    private @NotNull List<PathElement> convertToPathElements(int indentation, @NotNull String prefix,
-                                                             @NotNull List<String> elements) {
-        List<PathElement> pathElements = new ArrayList<>(elements.size());
-        for (String element : elements) {
-            List<String> comments = isFirstProperty
-                ? getCommentsIncludingRoot(prefix + element)
-                : configurationData.getCommentsForSection(prefix + element);
-            pathElements.add(new PathElement(indentation, element, comments, isFirstProperty));
+    public @NotNull List<PathElement> getPathElements(@NotNull String path) {
+        String[] pathParts = path.split("\\.");
+        int totalParts = pathParts.length;
+        int levelOfFirstNewPart = returnLevelOfFirstNewPathElement(path);
+
+        StringBuilder fullPathBuilder = new StringBuilder();
+        List<PathElement> pathElements = new ArrayList<>(totalParts);
+        int level = 0;
+        for (int i = 0; i < totalParts; ++i) {
+            fullPathBuilder.append(pathParts[i]);
+            PathElement element = new PathElement(level, pathParts[i], fullPathBuilder.toString(), isFirstProperty);
+            element.setEndOfPath(i == totalParts - 1);
+            element.setFirstOfGroup(levelOfFirstNewPart == level);
+            pathElements.add(element);
+
+            ++level;
+            fullPathBuilder.append(".");
             isFirstProperty = false;
-            prefix += element + ".";
-            ++indentation;
         }
-        pathElements.get(0).setFirstOfGroup(true);
+        lastPath = path;
         return pathElements;
     }
 
-    private @NotNull List<String> getCommentsIncludingRoot(@NotNull String path) {
-        List<String> rootComments = configurationData.getCommentsForSection("");
-        if ("".equals(path)) {
-            return rootComments;
+    /**
+     * Returns the hierarchy level of the highest path element that is being visited for the first time. For example,
+     * if we previously processed {@code config.datasource.mysql.type} and the given path is
+     * {@code config.datasource.driver.version}, then the level for the path element "driver" is returned (i.e. 2).
+     *
+     * @param path the new path
+     * @return the level of the first new path element
+     */
+    private int returnLevelOfFirstNewPathElement(@NotNull String path) {
```

Comment:
* Change to protected so anyone extending it could reuse it? Protected final?
* Is there a test for a root property?